File: cpp/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */
#include "leetcode.h"
// @lc code=start
class Solution
{
public:
  int lengthOfLongestSubstring(string s)
  {
    vector<int> dict(256, -1);
    int srt = -1;
    int len = 0;
    for (int i = 0; i < s.size(); i++) {
      if (dict[s[i]] > srt) {
        srt = dict[s[i]];
      }
      dict[s[i]] = i;
      len = max({len, i - srt});
    }
    return len;
  }
};
// @lc code=end

File: python/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        used = {}
        max_length = start = 0
        for i,c in enumerate(s):
            if c in used and start <= used[c]:
                start = used[c] + 1
            else:
                max_length = max(max_length, i-start+1)
            used[c]=i
        return max_length
# @lc code=end

